core/src/test/java/com/linecorp/armeria/client/limit/ConcurrencyLimitingClientTest.java

```diff
@@ -216,28 +243,93 @@ void testUnlimitedRequestWithFaultyDelegate() throws Exception {
+        final HttpRequest req2 = newReq();
+        final HttpResponseWriter actualRes1 = HttpResponse.streaming();
+        final HttpResponseWriter actualRes2 = HttpResponse.streaming();
 
-    private static void waitForEventLoop() {
-        eventLoop.get().submit(() -> { /* no-op */ }).syncUninterruptibly();
+        when(delegate.execute(ctx1, req1)).thenReturn(actualRes1);
+        when(delegate.execute(ctx2, req2)).thenReturn(actualRes2);
+
+        final ConcurrencyLimit concurrencyLimit = ConcurrencyLimit.builder(1)
+                                                                  .policy(requestContext -> false)
+                                                                  .build();
+
+        final ConcurrencyLimitingClient client =
+                newDecorator(concurrencyLimit).apply(delegate);
+
+        // both requests should be delegated immediately, creating no deferred response.
+        final HttpResponse res1 = client.execute(ctx1, req1);
+        final HttpResponse res2 = client.execute(ctx2, req2);
+
+        verify(delegate).execute(ctx1, req1);
+        assertThat(res1.isOpen()).isTrue();
+        assertThat(client.numActiveRequests()).isEqualTo(2);
+
+        verify(delegate).execute(ctx2, req2);
+        assertThat(res2.isOpen()).isTrue();
+        assertThat(client.numActiveRequests()).isEqualTo(2);
+
+        // Complete the response, leaving no active requests.
+        closeAndDrain(actualRes1, res1);
+        closeAndDrain(actualRes2, res2);
+        await().untilAsserted(() -> assertThat(client.numActiveRequests()).isZero());
     }
 
-    private static HttpRequest newReq() {
-        return HttpRequest.of(HttpMethod.GET, "/dummy");
+    @Test
+    void enforcesConcurrencyControlOnMultipleClients() throws Exception {
+        final ClientRequestContext ctx1 = newContext();
+        final ClientRequestContext ctx2 = newContext();
+        final HttpRequest req1 = newReq();
+        final HttpRequest req2 = newReq();
+        final HttpResponseWriter actualRes1 = HttpResponse.streaming();
+        final HttpResponseWriter actualRes2 = HttpResponse.streaming();
+
+        when(delegate.execute(ctx1, req1)).thenReturn(actualRes1);
+
+        final ConcurrencyLimit concurrencyLimit = ConcurrencyLimit.builder(1)
+                                                                  .timeoutMillis(500)
+                                                                  .policy(requestContext -> true)
+                                                                  .build();
+
+        final ConcurrencyLimitingClient clientOne =
+                newDecorator(concurrencyLimit).apply(delegate);
+
+        final ConcurrencyLimitingClient clientTwo =
+                newDecorator(concurrencyLimit).apply(delegate);
+
+        // Send two requests, where only the first one is delegated.
+        final HttpResponse res1 = clientOne.execute(ctx1, req1);
+        final HttpResponse res2 = clientTwo.execute(ctx2, req2);
```

Comment:
How about using `aggregate()` to simplify the consumption, instead of `NoopSubscriber`?

```java
final AggregatedHttpResponse res1 = clientOne.execute(ctx1, req1).aggregate();
final AggregatedHttpResponse res2 = clientOne.execute(ctx2, req2).aggregate();
await().untilAsserted(() -> {
    assertThatThrownBy(() -> res2.whenComplete().join())
            ...
    assertThat(res2.isOpen()).isFalse()
})
```